cmd/webhook-tester/main.go

```diff
@@ -20,193 +21,318 @@ import (
+		sig = signature([]byte(secret), []byte(body))
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("X-GitHub-Event", "workflow_job")
+	req.Header.Set("X-Hub-Signature-256", "sha256="+sig)
+	if *idToken != "" {
+		req.Header.Set("Authorization", "Bearer "+*idToken)
+	}
 
-		fmt.Printf("Response Status: %s\n", resp.Status)
-		fmt.Printf("Response Body: %s\n", string(body))
+	client := &http.Client{
+		Timeout: 30 * time.Second,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	log.Printf("Status: %s", resp.Status)
+	log.Printf("Body: %s", string(respBody))
+
+	if resp.StatusCode != *expectHTTPStatus {
+		return fmt.Errorf("expected status %d, but got %d", *expectHTTPStatus, resp.StatusCode)
+	}
 
-		if resp.StatusCode != tc.expectedStatusCode {
-			fmt.Fprintf(os.Stderr, "Test case %q failed: expected status code %d, got %d\n", tc.name, tc.expectedStatusCode, resp.StatusCode)
-			os.Exit(1)
+	log.Printf("Successfully received expected status code %d.", *expectHTTPStatus)
+
+	if *expectNoBuild {
+		if err := verifyNoBuildTriggered(ctx, *projectID, *runID); err != nil {
+			return fmt.Errorf("failed to verify that no build was triggered: %w", err)
 		}
-		fmt.Printf("Test case %q passed.\n\n", tc.name)
+		log.Printf("Successfully verified that no build was triggered.")
+		return nil
 	}
 
-	fmt.Println("All test cases passed.")
+	// Only run verification if the test was expected to succeed.
+	if *expectHTTPStatus == http.StatusOK {
+		log.Printf("Successfully sent webhook payload. Now verifying...")
 
-	fmt.Println("--- Running end-to-end test case: build verification ---")
-	if err := verifyBuildTriggered(*targetURL, validPayload, *secret); err != nil {
-		fmt.Fprintf(os.Stderr, "End-to-end test failed: %v\n", err)
-		os.Exit(1)
+		if *verifyRunner {
+			if err := verifyRunnerOnline(ctx, *githubOwner, *githubRepo, *runID); err != nil {
+				return fmt.Errorf("failed to verify runner: %w", err)
+			}
+			log.Printf("Successfully verified runner is online.")
+		} else {
+			if err := verifyBuildTriggered(ctx, *projectID, *runID); err != nil {
+				return fmt.Errorf("failed to verify build trigger: %w", err)
+			}
+			log.Printf("Successfully verified build trigger.")
+		}
 	}
-	fmt.Println("End-to-end test case passed.")
+	return nil
 }
 
-func verifyBuildTriggered(targetURL, payload, secret string) error {
-	signature, err := generateSignature([]byte(payload), secret)
+// verifyBuildTriggered polls GCP to check if a build was triggered with the correct tag.
+func verifyBuildTriggered(ctx context.Context, projectID, runID string) error {
+	const (
+		maxRetries = 10
+		delay      = 30 * time.Second
+	)
+
+	filter := fmt.Sprintf(`tags="e2e-run-id-%s"`, runID)
+
+	client, err := cloudbuild.NewClient(ctx)
```

Comment:
You could move this up a layer and pass the client in to the two methods that need it instead of creating separate ones.